src/primitive/measure.jl

```diff
@@ -1,31 +1,32 @@
-using YaoBase, YaoArrayRegister
+using YaoBase, YaoArrayRegister, Random
 export Measure, AllLocs, ComputationalBasis
 
 """
-    Measure{N, K, OT} <: PrimitiveBlock{N}
-    Measure(n::Int; operator=ComputationalBasis(), locs=1:n, collapseto=nothing, remove=false)
+    Measure{N, K, OT, RNG} <: PrimitiveBlock{N}
+    Measure(n::Int; rng=Random.GLOBAL_RNG, operator=ComputationalBasis(), locs=1:n, collapseto=nothing, remove=false)
 
 Measure operator.
 """
-mutable struct Measure{N, K, OT} <: PrimitiveBlock{N}
+mutable struct Measure{N, K, OT, RNG} <: PrimitiveBlock{N}
+    rng::RNG
     operator::OT
     locations::Union{NTuple{K, Int}, AllLocs}
     collapseto::Union{Int, Nothing}
     remove::Bool
     results::Vector{Int}
-    function Measure{N, K, OT}(operator, locations, collapseto, remove) where {N, K, OT}
+    function Measure{N, K, OT, RNG}(rng::RNG, operator, locations, collapseto, remove) where {RNG, N, K, OT}
```

Comment:
Can we give `operator`, `locs` and `collapseto`, `remove` a default value here? so both interface works. I still don't feel consistent to have that keyword only, but we can have both.

Reply:
For user, it is recommended to use the key word interface. The point is this parameter list is too long, it is hard to decide an order that make sense. I am considering moving `rng` in measure functions to kwargs, considering the little possibility of dispatching over it, and it is a marginal parameter.

Reply:
I don't really care about the dispatching, we should make it intuitive to non-Yao Julians. Or let's just let the original issue author say something? I prefer `rng` as the first argument. It's only about consistency. And even if we support kwargs, it is not a conflict, we can have both.

@zenna any suggestions?

Reply:
There are a few things to consider
- You can't dispatch on kwargs
- The convention in most of the packages that use randomness is to have `rng` as the first argument

In my own code (esp Omega.jl)  I've followed this convention.  It's a bit cumbersome to write as it normally means you have to have two definitions of things.  This is because the first argument can't have a default value if the following don't.

For example:

```julia
function Base.rand(rng::AbstractRNG,
                   x::RandVar,
                   n::Integer;
                   alg::SamplingAlgorithm = defalg(x),
                   ΩT::Type{OT} = defΩ(alg),
                   kwargs...) where {OT <: Ω}
  rand(rng, x, n, alg; ΩT = ΩT, kwargs...)
end

function Base.rand(x::RandVar,
                   n::Integer,
                   alg::SamplingAlgorithm;
                   ΩT::Type{OT} = defΩ(alg),
                   kwargs...) where {OT <: Ω}
  rand(Random.GLOBAL_RNG, x, n, alg; ΩT = ΩT, kwargs...)
end
```

Reply:
The convension of a Yao block is using nqubit as the first parameter, which is important for lazy construction. To measure, rng is a marginal parameter. To sample, it is not. If rng can not be the first parameter, I'd prefer it be a kwargs.